### ISSUE TITLE
Trouble with Balance Assertions/Assignments

### ISSUE THREAD

**Issue by OptikSage** [ledger/ledger#538]
Maybe I am missing something, but I seem to be having issues with balance assertions/assignments. I am trying to track my starbucks points. Sometimes points are earned through spending, and sometimes there are bonus points. My plan was to use automatic transactions for stars earned through spending and then just set the balance to zero (or whatever the balance is after redeeming) when the points are redeemed. Here is the test journal I have put together.

```
; -*- ledger -*-

account Awards:Starbucks:Earned
account Awards:Starbucks:Balance
account Awards:Starbucks:Redeemed

commodity $
    note United States Dollars
    format $1,000.00
    nomarket
    default
   
= @/^Starbucks/ and /Coffee/
    Awards:Starbucks:Earned                              (-0.5 SB * amount)
    Awards:Starbucks:Balance                             ( 0.5 SB * amount)

2017-01-01 * Starbucks (Card Reloaded)
    Expenses:Coffee                                      50.00 USD
    Liabilities:Cash                                       

2017-01-03 * Starbucks (Stars Redeemed)
    Awards:Starbucks:Earned                                -95 SB   ; ## Ultimately I want this auto calculated. ##
    Awards:Starbucks:Redeemed                             8.00 USD @@ 120.00 SB
    Awards:Starbucks:Balance
```

at this point this command give me the following (accurate) balance: $ ledger -f .\Awards.dat -w bal awards -V --empty --flat

                   0  Awards:Starbucks:Balance
           -120.0 SB  Awards:Starbucks:Earned
            120.0 SB  Awards:Starbucks:Redeemed

when I change the last transaction to 

```
2017-01-03 * Starbucks (Stars Redeemed)
    Awards:Starbucks:Redeemed                             8.00 USD @@ 120.00 SB
    Awards:Starbucks:Balance    
```                          

and add 

```
2017-01-03 * Starbucks (Stars Redeemed)
    Awards:Starbucks:Earned
    Awards:Starbucks:Balance                             = 0 SB 

```
I get the following balance:
```

$ ledger -f .\Awards.dat -w bal awards -V --empty --flat
             25.0 SB  Awards:Starbucks:Balance
           -145.0 SB  Awards:Starbucks:Earned
            120.0 SB  Awards:Starbucks:Redeemed
```

The question I have is why isn't Awards:Starbucks:Balance = 0 SB?  Isn't that what the balance assertion is supposed to insure? And if it does, then shouldn't it also set the last Awards:Starbucks:Earned posting to -95 SB and ultimately resulted in the same balance of

```
                   0  Awards:Starbucks:Balance
           -120.0 SB  Awards:Starbucks:Earned
            120.0 SB  Awards:Starbucks:Redeemed

```
Hope that made sense. Thank.

**Comment by jwiegley** [comment]
Your example file doesn't quite make sense to me, in terms of how Starbucks Stars work (something that I also partake in quite often). I would have expected a file that looked more like this:

```
; -*- ledger -*-

account Awards:Starbucks:Earned
account Awards:Starbucks:Balance
account Awards:Starbucks:Redeemed

commodity $
    note United States Dollars
    format $1,000.00
    nomarket
    default

= /^Expenses:Starbucks:/
    (Awards:Starbucks:Stars)                            (0.5 SB * amount)

2017-01-01 * Starbucks (Card Reloaded)
    Assets:Starbucks Card                                50.00 USD
    Liabilities:Cash

2017-01-02 * Starbucks Coffee
    Expenses:Starbucks:Coffee                             5.00 USD
    Assets:Starbucks Card

2017-01-03 * Starbucks (Stars Redeemed)
    (Awards:Starbucks:Stars)                            -120.00 SB
```

**Comment by jwiegley** [comment]
It looks like the increase caused by the automated transaction in this file is being applied *after* the balance assignment sets the account to 0, resulting in the account ultimately becoming `25 SB`. This is definitely a bug.

**Comment by jwiegley** [comment]
For future reference, here is the final data file showing the bug:

```
= @/^Starbucks/ and /Coffee/
    Awards:Starbucks:Earned                              (-0.5 SB * amount)
    Awards:Starbucks:Balance                             ( 0.5 SB * amount)

2017-01-01 * Starbucks (Card Reloaded)
    Expenses:Coffee                                      50.00 USD
    Liabilities:Cash

2017-01-03 * Starbucks (Stars Redeemed)
    Awards:Starbucks:Redeemed                             8.00 USD @@ 120.00 SB
    Awards:Starbucks:Balance

2017-01-03 * Starbucks (Stars Redeemed)
    (Awards:Starbucks:Balance)                             = 0 SB
```

And the bug can be seen by running:

```
ledger -f ledger.dat -B reg balance
```

**Comment by OptikSage** [comment]
Thanks for taking a look. I spent a few hours trying various permutations to make sure I understood how ledger worked, and we pretty sure this was a bug, but also thought I might also be going insane. :)

For now I can probably simplify how I am doing this, either skip tracking the actual star earnings on transaction by transaction basis as well as take a closer look at the format you proposed.

It terms of how we are thinking differently, by approach was to track the amount of stars earned via actual transactions (although I think I got the ratio backwards in my example), but that left a problem with all of the bonus stars (buy 3 drink in a row and get 50 more). Those I was just going to try and auto calculate with the balance adjustment since I new what the final balance would be after entering in my star redemption (which would be the only thing I needed to do by hand, if I pulled in the actual transaction from the bank).

Thanks.
